### PR TITLE
feat: allow cacheService to evict specific regions

### PR DESCRIPTION
**Pull request by LukasLohoff**
## Description

Allows cacheController / cacheService to evict specific regions. i.e.:

```bash
curl -k -v -X POST 'https://localhost/cache/evict?regions=layers,groups,files'
```

`CacheService::evictCacheRegions` can be called with multiple regions:

```java
@Autowired
CacheService cacheService;

cacheService.evictCacheRegions("layers", "topics", "groups");
```

@terrestris/devs Please review

## Pull request type

- [ ] Bugfix
- [x] Feature
- [ ] Dependency updates
- [ ] Tests
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

## Do you introduce a breaking change?

- [ ] Yes
- [x] No

## Checklist

- [x] I understand and agree that the changes in this PR will be licensed under the 
  [Apache Licence Version 2.0](https://github.com/terrestris/shogun/blob/main/LICENSE).
- [x] I have followed the [guidelines for contributing](https://github.com/terrestris/shogun/blob/main/CONTRIBUTING.md).
- [x] The proposed change fits to the content of the [code of conduct](https://github.com/terrestris/shogun/blob/main/CODE_OF_CONDUCT.md).
- [ ] I have added or updated tests and documentation, and the test suite passes (run `mvn test` locally).
- [ ] I have added a screenshot/screencast to illustrate the visual output of my update.

### CODE REVIEW

File: shogun-lib/src/main/java/de/terrestris/shogun/lib/controller/CacheController.java [hunk 1]
```diff
@@ -42,19 +45,23 @@ public class CacheController {
...
+            } else {
+                service.evictCacheRegions(regions.toArray(new String[]{}));
+                log.info("Successfully evicted cache regions {}", regions);
+            }
 
             return ResponseEntity.ok().build();
         } catch (Exception e) {
             log.error("Could not evict the cache: {}", e.getMessage());
-            log.trace("Full stack trace: {}", e);
+            log.trace("Full stack trace:", e);
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shogun/lib/controller/CacheController.java:64]:
```suggestion
            log.trace("Full stack trace: ", e);
```

File: shogun-lib/src/main/java/de/terrestris/shogun/lib/service/CacheService.java [hunk 2]
```diff
@@ -43,4 +47,23 @@ public void evictCache() throws Exception {
...
         cache.evictAllRegions();
     }
+
+    public void evictCacheRegions(String... region) throws Exception {
+        if (region == null) {
+            return;
+        }
+        if (entityManager == null) {
+            throw new Exception("Could not get the entity manager.");
+        } else {
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shogun/lib/service/CacheService.java:57]: `else` is not necessarily needed.

File: shogun-lib/src/main/java/de/terrestris/shogun/lib/service/CacheService.java [hunk 3]
```diff
@@ -43,4 +47,23 @@ public void evictCache() throws Exception {
...
+            EntityManagerFactory entityManagerFactory = entityManager.getEntityManagerFactory();
+            SessionFactory sessionFactory = entityManagerFactory.unwrap(SessionFactory.class);
+            Cache cache = sessionFactory.getCache();
+            for (String r : region) {
+                if (StringUtils.isEmpty(r)) {
+                    continue;
+                }
+                cache.evictRegion(r);
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shogun/lib/service/CacheService.java:65]: What happens for an unavailable region identifier? Do we need to check if the region exists before?

**Review by dnlkoch — APPROVED** [review]
This is a very useful addition, thanks @LukasLohoff!

I added just some minor remarks and a question regarding the (required?) check for the region existence. If not needed, just ignore.

File: shogun-lib/src/main/java/de/terrestris/shogun/lib/service/CacheService.java (hunk 2 above)
Comment (LukasLohoff) [reply shogun-lib/src/main/java/de/terrestris/shogun/lib/service/CacheService.java]: indeed! removed in e2e58ba45085a61b7588768bf83814757a683e13

File: shogun-lib/src/main/java/de/terrestris/shogun/lib/service/CacheService.java (hunk 3 above)
Comment (LukasLohoff) [reply shogun-lib/src/main/java/de/terrestris/shogun/lib/service/CacheService.java]: good point. I'll do some testing

Comment (LukasLohoff) [reply shogun-lib/src/main/java/de/terrestris/shogun/lib/service/CacheService.java]: A NullPointerException is thrown if one region identifier is invalid, preventing the following regions from being cleared. AFAIK there is no hibernate method which checks if an identifier is valid or not.

I could wrap `cache.evictRegion(r);` into a try-catch block. That way, other (possible) valid regions will still be cleared.

Comment (LukasLohoff) [reply shogun-lib/src/main/java/de/terrestris/shogun/lib/service/CacheService.java:66]: implemented here: 9cfa96669971e5da3e06bdba9a5a1270dd3bc7b2

**Review by dnlkoch — APPROVED** [review]

**Pull request merged**